src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml

```diff
@@ -381,66 +347,123 @@ and gen_from_arrow ~loc ~env left right =
     | x -> (gen_from_type ~loc ~env x, [%expr o_nil])
   in
   let x, obs = aux right in
+  (* TODO: export this in qcheck_generators for https://github.com/c-cube/qcheck/issues/190 *)
   let arb = [%expr QCheck.make [%e x]] in
   [%expr
     QCheck.fun_nary QCheck.Tuple.([%e observable left] @-> [%e obs]) [%e arb]
     |> QCheck.gen]
 
-let gen_from_kind_variant ~loc ~env typ_name xs =
-  let (module A) = Ast_builder.make loc in
-  let is_rec (constr : constructor_declaration) : bool =
-    match constr.pcd_args with
-    | Pcstr_tuple xs -> List.exists (is_rec_typ typ_name) xs
-    | _ -> false
-  in
-  sized ~loc ~env typ_name is_rec (gen_from_constr ~loc) xs
+(** [gen_from_type_declaration loc td] creates a generator from the type declaration.
+
+    It returns either `Recursive or `Normal.
 
-let rec curry_args ~loc args body =
-  match args with
-  | [] -> body
-  | x :: xs -> [%expr fun [%p x] -> [%e curry_args ~loc xs body]]
+    - `Normal of expression:
+    The derived generator is not recursive, we return only the generator.
 
-let gen_from_type_declaration ~loc ?(env = TypeGen.empty) td =
-  let name = td.ptype_name.txt in
-  let pat_gen = pat ~loc name in
+    - `Recursive of expression * expression
+    The derived generator was recursive (i.e. val gen : n -> t Gen.t), we return
+    the sized generator version, and a normal generator using this last with
+    [Gen.sized].
+*)
+let gen_from_type_declaration ~loc ~env td =
+  let (module A) = Ast_builder.make loc in
+  let ty = env.Env.curr_type in
+  let is_rec = Env.is_rec env ty in
 
   let args =
     List.map
       (fun (typ, _) ->
-        match typ.ptyp_desc with Ptyp_var s -> pat ~loc s | _ -> assert false)
+        match typ.ptyp_desc with
+        | Ptyp_var s -> (pat ~loc s, name s |> A.evar)
+        | _ -> assert false)
       td.ptype_params
   in
+  let (args_pat, args_expr) = List.split args in
 
   let gen =
     match td.ptype_kind with
-    | Ptype_variant xs -> gen_from_kind_variant ~loc ~env name xs
+    | Ptype_variant xs ->
+       let is_rec cd = is_rec_constr_decl env cd in
```

Comment:
Just spotted  a different indentation than the following two cases

Reply:
Done, thank you for catching this